src/ball_sort/solver.cpp

```diff
@@ -0,0 +1,118 @@
+#include "ball_sort/solver.hpp"
+#include "ball_sort/move.hpp"
+#include "ball_sort/tube.hpp"
+#include <iostream>
+
+auto Solver::solve(Puzzle &puzzle) -> void
+{
+    puzzle.reset();
+
+    print_puzzle(puzzle);
+
+    std::unordered_set<Move> excluded_moves{};
```

Comment:
Seems like my version of Clang doesn't support flat_set.

I ran some tests in a separate sandbox with a vector and an unordered_set of 10000 items (reserved).

Things that I learnt:
- reserving is good
- vectors smash sets for emplacing
- std::find is better than std::unordered_set::contains, although I don't know exactly why yet
- sets smash vectors for lookup

Nothing particularly surprising there.

However, for some reason when I actually run the program with excluded_moves as a vector, it takes almost twice as long (120ms vs 70ms).

I assume this is because I'm doing more lookups than emplaces.

Reserving the size of excluded moves makes a fractional improvement to the vector performance.

Applying compiler optimisations makes them equivalent at around 30ms, though.

I'm inclined to go with the set for now, since it's faster without optimisation and equivalent with. But this has been a fun learning journey.

Reply:
Actually, std::unordered set is faster with optimisation too. 20ms vs 30ms for the vector.

That seals it.

Reply:
18ms with reserve. Woop.